components/proxy/src/test/kotlin/com/hotels/styx/services/OriginsAdminHandlerTest.kt

```diff
@@ -0,0 +1,205 @@
+/*
+  Copyright (C) 2013-2019 Expedia Inc.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+package com.hotels.styx.services
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.hotels.styx.*
+import com.hotels.styx.ErrorResponse
+import com.hotels.styx.api.HttpRequest
+import com.hotels.styx.api.HttpResponseStatus
+import com.hotels.styx.api.HttpResponseStatus.*
+import com.hotels.styx.api.LiveHttpRequest
+import com.hotels.styx.infrastructure.configuration.json.mixins.ErrorResponseMixin
+import com.hotels.styx.routing.RoutingMetadataDecorator
+import com.hotels.styx.routing.RoutingObjectRecord
+import com.hotels.styx.routing.db.StyxObjectStore
+import com.hotels.styx.routing.mockObject
+import com.hotels.styx.server.HttpInterceptorContext
+import io.kotlintest.matchers.collections.shouldContain
+import io.kotlintest.matchers.string.shouldStartWith
+import io.kotlintest.shouldBe
+import io.kotlintest.specs.FeatureSpec
+import io.mockk.mockk
+import reactor.core.publisher.Mono
+import java.nio.charset.StandardCharsets.UTF_8
+
+class OriginsAdminHandlerTest : FeatureSpec({
+
+    val mapper = ObjectMapper().addMixIn(ErrorResponse::class.java, ErrorResponseMixin::class.java)
+
+    val store = StyxObjectStore<RoutingObjectRecord>()
+    val mockObject = RoutingMetadataDecorator(mockObject())
+    val handler = OriginsAdminHandler("/base/path", "testProvider", store)
+
+    store.insert("app.active", RoutingObjectRecord("HostProxy", setOf(sourceTag("testProvider"), stateTag(STATE_ACTIVE)), mockk(), mockObject))
+    store.insert("app.closed", RoutingObjectRecord("HostProxy", setOf(sourceTag("testProvider"), stateTag(STATE_INACTIVE)), mockk(), mockObject))
+    store.insert("app.unreachable", RoutingObjectRecord("HostProxy", setOf(sourceTag("testProvider"), stateTag(STATE_UNREACHABLE)), mockk(), mockObject))
+    store.insert("app.nothostproxy", RoutingObjectRecord("NotHostProxy", setOf(sourceTag("testProvider"), stateTag(STATE_UNREACHABLE)), mockk(), mockObject))
+
+    fun getResponse(handler: OriginsAdminHandler, request: LiveHttpRequest) =
+            Mono.from(Mono.from(handler.handle(request, HttpInterceptorContext.create())).block().aggregate(1000)).block()
+
```

Comment:
Unnecessary mono & `.block()`. I'd do this:

```
fun Eventual<LiveHttpResponse>.wait(maxBytes: Int = 100*1024, debug: Boolean = false) = this.toMono()
        .flatMap { it.aggregate(maxBytes).toMono() }
        .doOnNext {
            if (debug) {
                println("${it.status()} - ${it.headers()} - ${it.bodyAs(StandardCharsets.UTF_8)}")
            }
        }
        .block()
```

Reply:
I see. You've replaced the first `block()` with a `.flatMap()` to process the aggregated `HttpResponse`, so you only need the final `block()` - there's no need to wait twice.